File: insertion_sort.py
def insertionSort(array):
    for i in range(1, len(array)):
        j = i
        while j > 0 and array[j-1] > array[j]:
            swap(j-1, j, array) 
            j -= 1
    return array

def swap(i, j, array):
    array[i], array[j] = array[j], array[i]


if __name__ == "__main__":
    array = [8, 2, 5, 4, 1]
    print(insertionSort(array))

"""
output: [1, 2, 4, 5, 8]
"""

File: insertion_sort.go
package main

import "fmt"

func insertionSort(array []int) []int {
	for i := 1; i < len(array); i++ {
		j := i
		for j > 0 && array[j-1] > array[j] {
			swap(j-1, j, array)
			j -= 1
		}
	}
	return array
}

func swap(i, j int, array []int) {
	array[i], array[j] = array[j], array[i]
}

func main() {
	array := []int{8, 2, 5, 4, 1}
	fmt.Println(insertionSort(array))
}

/* output
[1 2 4 5 8]
*/
